### PR TITLE
[luci-interpreter] Introduce Pack kernel

### PR DESCRIPTION
**Pull request by struss**
For #6262

This commit will introduce `Pack` kernel to enable evaluation `CirclePack`.

ONE-DCO-1.0-Signed-off-by: KiDeuk Bang <rrstrous@nate.com>

### CODE REVIEW

**Comment by struss** [comment]
@nnfw-bot test nncc-release

**Review by seanshpark — APPROVED** [review]
LGTM

File: compiler/luci-interpreter/src/kernels/Pack.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+{
+}
+
+void Pack::configure()
+{
+  LUCI_INTERPRETER_CHECK(_inputs.size() == static_cast<uint32_t>(params().values_count));
+  const Tensor *t0 = _inputs[0];
+  const int dimension_size = t0->shape().num_dims() + 1;
+  int axis = params().axis;
+  while (axis < 0)
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Pack.cpp:41]:
```suggestion
  if (axis < 0)
```

File: compiler/luci-interpreter/src/kernels/Pack.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  {
+    axis += dimension_size;
+  }
+  LUCI_INTERPRETER_CHECK(axis >= 0 && axis < t0->shape().num_dims());
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Pack.cpp:45]:
```suggestion
  LUCI_INTERPRETER_CHECK(axis >= 0 && axis <= t0->shape().num_dims());
```

File: compiler/luci-interpreter/src/kernels/Pack.cpp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Pack.cpp:46]: Please add checks for the below code

From tensorflow code
```cpp
  if (input0->type != kTfLiteInt32 && input0->type != kTfLiteFloat32 &&
      input0->type != kTfLiteUInt8 && input0->type != kTfLiteInt8 &&
      input0->type != kTfLiteInt16 && input0->type != kTfLiteInt64) {
    context->ReportError(context, "Type '%s' is not supported by pack.",
                         TfLiteTypeGetName(input0->type));
    return kTfLiteError;
  }
```

File: compiler/luci-interpreter/src/kernels/Pack.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    const Tensor *tensor = _inputs[i];
+    LUCI_INTERPRETER_CHECK(tensor->element_type() == t0->element_type());
+    LUCI_INTERPRETER_CHECK(tensor->shape().num_dims() == t0->shape().num_dims());
+    for (int d = 0; d < t0->shape().num_dims(); ++d)
+    {
+      LUCI_INTERPRETER_CHECK(tensor->shape().dim(d) == t0->shape().dim(d));
+    }
+  }
+
+  Shape output_shape = Shape(dimension_size);
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Pack.cpp:58]:
```suggestion
  Shape output_shape(dimension_size);
```

File: compiler/luci-interpreter/src/kernels/Pack.cpp (hunk 1 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/Pack.cpp:41]: Okay I will fix it. Thanks!

File: compiler/luci-interpreter/src/kernels/Pack.cpp (hunk 3 above)
Comment (jinevening) [reply compiler/luci-interpreter/src/kernels/Pack.cpp:46]: Never mind. It is done in `execute`.

File: compiler/luci-interpreter/src/kernels/Pack.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    {
+      output_shape.dim(index) = params().values_count;
+    }
+    else
+    {
+      output_shape.dim(index) = t0->shape().dim(i++);
+    }
+  }
+
+  if (t0->quantized_dimension() != 0)
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Pack.cpp:72]: There can be a quantized tensor whose `quantized_dimension` is 0.

If you want to check t0 is quantized or not, you should check if the type of t0 is U8, S8, S16, S32, or S64.

File: compiler/luci-interpreter/src/kernels/Pack.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      throw std::runtime_error("Unsupported type.");
+  }
+}
+
+template <typename T> void Pack::evalGeneric() const
+{
+  const Tensor *t0 = _inputs[0];
+  const int dimension_size = t0->shape().num_dims() + 1;
+  int axis = params().axis;
+  while (axis < 0)
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Pack.cpp:120]:
```suggestion
  if (axis < 0)
```

File: compiler/luci-interpreter/src/kernels/Pack.test.cpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  Check<TypeParam>(/*input_shapes=*/{{2}, {2}, {2}},
+                   /*output_shape=*/{3, 2},
+                   /*input_datas=*/
+                   {{1, 4}, {2, 5}, {3, 6}},
+                   /*output_data=*/
+                   {1, 4, 2, 5, 3, 6}, 0);
+
+  SUCCEED();
+}
+
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Pack.test.cpp:76]: Can you add a test for negative axis?

File: compiler/luci-interpreter/src/loader/KernelBuilder.cpp [hunk 8]
```diff
@@ -626,6 +627,22 @@ std::unique_ptr<Kernel> KernelBuilder::visit(const luci::CircleOutput *)
   throw std::runtime_error("Output node cannot be executed.");
 }
 
+std::unique_ptr<Kernel> KernelBuilder::visit(const luci::CirclePack *node)
+{
```
Comment (jinevening) [inline compiler/luci-interpreter/src/loader/KernelBuilder.cpp:631]: Can you add the assert?

`assert(node->arity() == node->values_count());`

File: compiler/luci-interpreter/src/kernels/Pack.cpp (hunk 2 above)
Comment (jinevening) [reply compiler/luci-interpreter/src/kernels/Pack.cpp:45]: Please ignore the comment. 

This change was from TFLite, but I think your code is right.

Comment (struss) [reply compiler/luci-interpreter/src/kernels/Pack.cpp:45]: Okay I will fix it. Thanks!

File: compiler/luci-interpreter/src/kernels/Pack.cpp (hunk 4 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/Pack.cpp:58]: Okay I will fix it. Thanks!

File: compiler/luci-interpreter/src/kernels/Pack.test.cpp (hunk 7 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/Pack.test.cpp:76]: Okay I will add.

File: compiler/luci-interpreter/src/kernels/Pack.cpp (hunk 6 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/Pack.cpp:120]: I will fix it.

File: compiler/luci-interpreter/src/loader/KernelBuilder.cpp (hunk 8 above)
Comment (struss) [reply compiler/luci-interpreter/src/loader/KernelBuilder.cpp:631]: I will add it.

File: compiler/luci-interpreter/src/kernels/Pack.cpp (hunk 3 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/Pack.cpp:46]: okay I will update.

**Review by jinevening — APPROVED** [review]
LGTM Thank you!

**Review by seanshpark — APPROVED** [review]
LGTM
